sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -63,4 +63,6 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
 case object StringType extends StringType(0) {
   val DEFAULT_COLLATION_ID = 0
   def apply(collationId: Int): StringType = new StringType(collationId)
+
+  def unapply(s: StringType): Option[Int] = Some(s.collationId)
```

Comment:
Let's not do this (I already tried this but was stopped by @cloud-fan). All the changes in this code path are very touchy, given that this is part of publicly exposed API. Plus, AFAIK you don't even need this change, given that this is case object?